accepted/future-releases/language-versioning/package-config-file-v2.md

```diff
@@ -0,0 +1,134 @@
+# Dart Package Configuration File v2.0
+
+Author: lrn@google.com
+Version: 0.1
+
+Dart currently uses a `.packages` file to configure *package URI resolution* (converting a package URI to a file name). The file uses a `.ini` file-like format with package names as keys and URI references as values. The URI references generated by Pub are always absolute or relative paths, so third-party users of the file tend to see them as paths, not URI.
+
+For the *language versioning* feature, we have proposed extending the format by adding fragments to the URI references and an entry with the empty string as key. This is a breaking change (the empty key was not previously accepted) and even more breaking for tools which assume the values are paths.
+
+Instead of breaking the existing file users, we will instead introduce a *new* file containing the information that we intended to store in the `.packages` file. This fill will use a (more) structured format which allows extending it with further information in the future. We will still generate the `.packages` file for now, allowing users of the file to continue working until they can migrate to using the new file.
+
+## New File Contents
+
+The file will contain the information that was planned for the `.packages` file:
+
+- A *default package* specified by its package name.
+- A list of *packages*, each with:
+  - A *package name*.
+  - A *path* or *location URI*.
+  - A *language version* (major/minor version number)
+
+On top of that, we also add extra metadata which allows us to version the format and attach information about when and how the file was created.
+
+We use JSON for the format. JSON has parsers available for most languages (not all Dart tools are written in Dart) and it is standardized and efficiently parsable.  JSON is *not* easy to write by hand (but doable), and it does not allow comments, but this is a machine generated file that will hardly ever be edited by hand.
+
+## New File Format
+
+The new file will be called `.packages.json` and stored in the same directory as the `.packages` file. The format is JSON text describing is a single JSON object.
+
+That object has, at least, the following properties:
+
+- `apiVersion`: A single integer marking the version of this format, currently `2` (counting `.packages` as the first). Tools should refuse to work on files with a larger version number than the ones they are aware of. Incremental changes will not require increasing the version, so new versions are not expected any time soon, but code should still prepare for it.
+
+- `defaultPackage`: A string containing a package name.
```

Comment:
I think this refers to the same concept that we sometimes call "root package" in the build system and docs. Can we consider using that name here? Can we expand this description to explain exactly what this means and how tools should interpret it?

Reply:
I don't like "root package" since I don't know what it means (there is no tree structure, so why is it a root). It might commonly be used for the "current" Pub package, but this file is ignorant of Pub (it will be generated by other tools too). 

This is the package to use for files which do not otherwise belong to a package. It allows us to attach properties to non-package files which we otherwise attach to packages.

Reply:
> I don't like "root package" since I don't know what it means (there is no tree structure, so why is it a root). 

It is the root of the dependency graph. It is the package that contains the entrypoint being invoked, which in turn imports all other libraries used by the application. I think "root" is a reasonable word for that.

"Default" implies that the implementation needs to "select" a package for some purpose and in the absence of other information it selects this one. I guess that is your mindset — that stuff outside of "lib" is not in any package so when it needs to guess at the package for it, it picks the default. But I don't think that's how any significant fraction of users look at the system. To users, all files under a package's root directory, even outside of "lib" are clearly and explicitly part of that package.

For example, given:

```
foo/
  lib/
    foo.dart
  example/
    foo_app.dart
bar/
  lib/
    bar.dart
  example/
    bar_app.dart
```

And a packages.json file like:

```json
{
  "defaultPackage": "foo",
  "packages": [
    {
      "name": "foo",
      "path": "foo/"
    },
    {
      "name": "bar",
      "path: "bar/"
    }
  ]
}
```

If a user somehow contrived to invoke `bar/example/bar_app.dart` using that
"packages.json" file, they would be very surprised if "bar.dart" was interpreted at the language version specified by *foo*.

You might argue that they are only surprised because they don't understand how the underlying model works. But I think the reality is that the underyling model where "package" means something different from the user-facing notion of "package" they have been taught for year is misleading. We shouldn't declare that their brains are wrong. We should design the system to be consistent with what's in their brains.

Maybe "default package" is the wrong notion entirely. All that does is determine the language version, right? How about we state that explicitly? Something like:

```json
{
  "dart": {
    "version": "2.1",
  },
  "packages": [
    {
      "name": "foo",
      "path": "foo/"
    },
    {
      "name": "bar",
      "path: "bar/"
    }
  ]
}
```

?

Reply:
The issue here is that there are different models in play: The Dart package and the Pub package.

The Dart language has no notion of root package or dependency graphs. It has a finite set of available Dart packages, which each contains a hierarchical structure of available Dart files that you can find using `package:` URIs. Then you run one of the available scripts to start a program. You can start more scripts, from different packages, using `Isolate.spawnUri`. Tree shaking is a compiler optimization that ensures that your deployed code only contains what is reachable from an entry point.

The Pub package is more than that, but it morally belongs to the same Pub package, and since that is what users are thinking about when they develop a package, the distinction between a Pub package and a Dart package is confusing. 

The Pub package concept is *so far* not reflected in the language or compiler tools, which is not surprising since it was developed *after* the language.

*We could change this now*. We could say that the `package_config.json` file will point to a *Pub* package directory (not its `lib/` dir), package URI resolution would use the `lib/` dir, but language versioning would apply the language version to all files in the Pub package directory tree.
I think that would be compatible with the current Pub cache layout (it contains `lib` dirs) and the Google Dart code layout (a Google Dart package has a root dir and a `lib/` dir as well).

It would hard-code the Pub layout (at least the `lib/` sub-dir) into tools which are currently agnostic to how packages are stored.

Or we could make the "lib" path configurable and have package entries like:
```json
{
  "name": "foo",
  "rootUri": "../", /* relative to JSON file */
  "libUri": "lib/", /* relative to rootUri */
  "languageVersion": "2.5"
}
```
where the `rootUri` is what we use to determine whether a file belongs to the package (is it in a subdir of this location, then it belongs to the package, and if two packages overlap, we'll resolve it somehow, perhaps by flipping a coin), and the `libUri` is a directory that we resolve `package:` URIs against (and which *must* be a subdir of the `rootUri`, no cheating!). That allows us to expand the "belongs to a package" to a directory tree (heck, we can add "includes" and "excludes" for subdirs as well if we want to, and do fancy Boolean logic to decide whether a file is "in the package" or not), and not just to the subdirectory where the Dart package files reside.

Or we can say that if `libUri` is absent, just use `rootUri`, and make `rootUri` point at the `lib/` directory for all other packages, and only care about the extra files of the current package. Pub can do what it wants here, this will give it tools to say what the "Pub package" is, and not just the "Dart package" (but it also has to say that, obviously, we do need to resolve `package:` URIs).

Should we do something like this?


(I added the "default package" because I want to allow the files in a Pub package to be considered part of the package for other (potential) future features, like package privacy or package sealed types. It was the *minimal* feature which allows one (*current*) package to have its test/ and other directories be considered in same package when you run your tests. The suggestion here would satisfy that, while also allowing a more general use. I just want to make absolutely sure that nobody will be in a position to claim that their own code is somehow part of a third-party package to get access to package private code.)

Reply:
> The Dart language has no notion of root package or dependency graphs.

For a given program invocation, there is a single library that is the entrypoint where execution began. All other libraries are reached from the import/export graph starting at that library. That sounds like a root package and a dependency graph to me.

> The Pub package concept is _so far_ not reflected in the language or compiler tools, which is not surprising since it was developed _after_ the language.

Well, *sort of*. The "package:" import scheme was added to the language by Kasper et al. specifically at the behest of Natalie and I because we were in the middle of building pub and needed it. Dart existed before pub, but the language's notion of packages did not.

> _We could change this now_. We could say that the `package_config.json` file will point to a _Pub_ package directory (not its `lib/` dir), package URI resolution would use the `lib/` dir, but language versioning would apply the language version to all files in the Pub package directory tree.

Not a bad idea. The fact that the language tries to define its own somewhat different notion of package has never seemed practically useful to me.

> It would hard-code the Pub layout (at least the `lib/` sub-dir) into tools which are currently agnostic to how packages are stored.

I'm a little hesitant to bake directory policy right into the language. We agonized over how "package:" imports should map to the interior of pub packages for a long time before settling on "lib" in large part to avoid symlink cycles. It's a *useful* policy, and I think a pretty good one, but I don't know if we want to be stuck with it forever.

> the `libUri` is a directory that we resolve `package:` URIs against (and which _must_ be a subdir of the `rootUri`, no cheating!).

Interesting idea! I do really like the idea of having the language's notion of "package" include the entire Pub package directory. Like you note, that may come into play with things like package privacy. It would be *really* nice if, for example, tests could have some kind of privileged access to things in the same package's libraries.

It might be worth thinking about a little more but my initial feeling is that I like it.

Reply:
So I did it.